File: mobile/src/main/java/com/siliconlabs/bledemo/adapters/LogAdapter.java
package com.siliconlabs.bledemo.adapters;

import android.content.Context;

import androidx.annotation.NonNull;
import androidx.core.content.ContextCompat;
import androidx.recyclerview.widget.RecyclerView;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import com.siliconlabs.bledemo.other.LogType;
import com.siliconlabs.bledemo.R;
import com.siliconlabs.bledemo.log.Log;

import java.util.List;

public class LogAdapter extends RecyclerView.Adapter<LogAdapter.LogViewHolder> {
    private List<Log> logList;
    private Context context;
    private String deviceAddress = null;
    private boolean logByAddress = false;

    public class LogViewHolder extends RecyclerView.ViewHolder {
        TextView timeTV;
        TextView infoTV;

        public LogViewHolder(View itemView) {
            super(itemView);
            timeTV = itemView.findViewById(R.id.textview_log_time);
            infoTV = itemView.findViewById(R.id.textview_log_info);

        }
    }

    public LogAdapter(List logList, Context context) {
        this.logList = logList;
        this.context = context;
    }

    public void logByDeviceAddress(String deviceAddress) {
        this.deviceAddress = deviceAddress;
        logByAddress = true;
        notifyDataSetChanged();
    }


    @Override
    public void onBindViewHolder(LogAdapter.LogViewHolder holder, int position) {
        Log log = logList.get(position);


        if (logByAddress) {
            if (log.getDeviceAddress().equals(this.deviceAddress)) {
                if (logList.get(position).getLogType().equals(LogType.CALLBACK)) {
                    holder.infoTV.setText("(Callback)" + log.getLogInfo());
                    holder.infoTV.setTextColor(ContextCompat.getColor(context, R.color.silabs_blue));
                } else {
                    holder.infoTV.setText(log.getLogInfo());
                    holder.infoTV.setTextColor(ContextCompat.getColor(context, R.color.silabs_subtle_text));
                }

                holder.timeTV.setText(logList.get(position).getLogTime());
            }
        } else {
            if (logList.get(position).getLogType().equals(LogType.CALLBACK)) {
                holder.infoTV.setText("(Callback)" + log.getLogInfo());
                holder.infoTV.setTextColor(ContextCompat.getColor(context, R.color.silabs_blue));
            } else {
                holder.infoTV.setText(log.getLogInfo());
                holder.infoTV.setTextColor(ContextCompat.getColor(context, R.color.silabs_subtle_text));
            }

            holder.timeTV.setText(logList.get(position).getLogTime());
        }
    }

    @Override
    public LogAdapter.LogViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.log_item, parent, false);
        return new LogViewHolder(view);
    }

    @Override
    public int getItemCount() {
        return logList.size();
    }

    public void setLogList(List<Log> logList) {
        this.logList = logList;
    }
}

File: mobile/src/main/java/com/siliconlabs/bledemo/Browser/Adapters/LogAdapter.kt
package com.siliconlabs.bledemo.Browser.Adapters

import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.core.content.ContextCompat
import androidx.recyclerview.widget.RecyclerView
import com.siliconlabs.bledemo.Browser.Adapters.LogAdapter.LogViewHolder
import com.siliconlabs.bledemo.Browser.Models.LogType
import com.siliconlabs.bledemo.Browser.Models.Logs.Log
import com.siliconlabs.bledemo.R

class LogAdapter(private var logList: List<Log>, private val context: Context) : RecyclerView.Adapter<LogViewHolder>() {
    private var deviceAddress: String? = null
    private var logByAddress = false

    inner class LogViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        private var tvTime = itemView.findViewById(R.id.tv_log_time) as TextView
        private var tvInfo = itemView.findViewById(R.id.tv_log_info) as TextView

        fun bind(log: Log) {
            if (logByAddress) {
                if (log.deviceAddress == deviceAddress) {
                    if (log.logType == LogType.CALLBACK) {
                        setCallbackLog(log)
                    } else {
                        setCommonLog(log)
                    }
                    tvTime.text = log.logTime
                }
            } else {
                if (log.logType == LogType.CALLBACK) {
                    setCallbackLog(log)
                } else {
                    setCommonLog(log)
                }
                tvTime.text = log.logTime
            }
        }

        private fun setCallbackLog(log: Log) {
            tvInfo.text = "(Callback)" + log.logInfo
            tvInfo.setTextColor(ContextCompat.getColor(context, R.color.silabs_blue))
        }

        private fun setCommonLog(log: Log) {
            tvInfo.text = log.logInfo
            tvInfo.setTextColor(ContextCompat.getColor(context, R.color.silabs_subtle_text))
        }
    }

    fun logByDeviceAddress(deviceAddress: String?) {
        this.deviceAddress = deviceAddress
        logByAddress = true
        notifyDataSetChanged()
    }

    override fun onBindViewHolder(holder: LogViewHolder, position: Int) {
        val log = logList[position]
        holder.bind(log)
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): LogViewHolder {
        val view = LayoutInflater.from(parent.context).inflate(R.layout.adapter_log, parent, false)
        return LogViewHolder(view)
    }

    override fun getItemCount(): Int {
        return logList.size
    }

    fun setLogList(logList: List<Log>) {
        this.logList = logList
    }
}
